golang/cosmos/ante/inbound.go

```diff
@@ -10,83 +10,114 @@ import (
-// in the action queue when executed.
+// NewInboundDecorator returns an AnteDecorator which honors the allowed size of the inbound queue.
 func NewInboundDecorator(sk SwingsetKeeper) sdk.AnteDecorator {
 	return inboundAnte{sk: sk}
 }
 
-// AnteHandle implements sdk.AnteDecorator
+// AnteHandle implements sdk.AnteDecorator.
+// Lazily consults the Swingset state to avoid overhead when dealing
+// with pure Cosmos-level Txs.
 func (ia inboundAnte) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
 	msgs := tx.GetMsgs()
 	inboundsAllowed := int32(-1)
 	for _, msg := range msgs {
-		switch msg.(type) {
-		case *swingtypes.MsgDeliverInbound,
-			*swingtypes.MsgInstallBundle,
-			*swingtypes.MsgProvision,
-			*swingtypes.MsgWalletAction,
-			*swingtypes.MsgWalletSpendAction:
-			// Lazily compute the number of allowed messages when the transaction
-			// includes a SwingSet message. This number is the difference between
-			// the number of allowed messages initially returned by SwingSet, and
-			// the number of messages already added into the actionQueue by other
-			// transactions included in the block.
-			// We store the number of allowed messages locally since messages are
-			// added to the actionQueue after going through the ante handler (in
-			// CheckTx) and before the next transaction is processed.
-			// However in CheckTx (mempool admission check), no state is
-			// changed so it's possible for a set of transactions to exist which
-			// if all included in a block would push the actionQueue over, and thus
-			// end up in rejections instead of simply not admitting them in the
-			// mempool. To mitigate this, Swingset should compute the number of
-			// messages allowed for mempool admission as if its max queue length
-			// was lower (e.g. 50%). This is the QueueInboundMempool entry.
-			if inboundsAllowed == -1 {
-				state := ia.sk.GetState(ctx)
-				entry := swingtypes.QueueInbound
-				if ctx.IsCheckTx() {
-					entry = swingtypes.QueueInboundMempool
-				}
-				allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, entry)
-				if found {
-					actions, err := ia.sk.ActionQueueLength(ctx)
-					if err != nil {
-						return ctx, err
-					}
-					if actions < allowed {
-						inboundsAllowed = allowed - actions
-					} else {
-						inboundsAllowed = 0
-					}
-				} else {
-					// if number of allowed entries not given, fail closed
-					inboundsAllowed = 0
-				}
-			}
-			if inboundsAllowed > 0 {
-				inboundsAllowed--
-			} else {
-				defer func() {
-					telemetry.IncrCounterWithLabels(
-						[]string{"tx", "ante", "inbound_not_allowed"},
-						1,
-						[]metrics.Label{
-							telemetry.NewLabel("msg", sdk.MsgTypeURL(msg)),
-						},
-					)
-				}()
-				return ctx, ErrInboundQueueFull
+		inbounds := inboundMessages(msg)
+		if inbounds == 0 {
+			continue
+		}
+		if inboundsAllowed == -1 {
+			var err error
+			inboundsAllowed, err = ia.allowedInbound(ctx)
```

Comment:
stylistic gloang question while I learn, why not do the following?
```suggestion
			inboundsAllowed, err := ia.allowedInbound(ctx)
```

Reply:
This is an ugly Go gotcha. In a [short variable declaration](https://go.dev/ref/spec#Short_variable_declarations) (_vars_ `:=` _exprs_) where at least one of the variables is new, any variables that have been previously declared *in the same block* with the same type will be treated as _redeclarations_. But if the previous declaration was in an enclosing block, it creates a new variable which shadows the previous variable.

I wish I could claim that I instinctively used the correct form, but in this case I was alerted to the shadowing by an unused variable warning. You can also use the `go vet` tool to check for common mistake patterns.

Here's a good list of [common Go gotchas](http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/).